rust/flexbuffers/src/builder/mod.rs

```diff
@@ -0,0 +1,370 @@
+        }
+    }
+}
+macro_rules! push_indirect {
+    ($push_name: ident, $scalar: ty, $Direct: ident, $Indirect: ident) => {
+        fn $push_name<T: Into<$scalar>>(&mut self, x: T) {
+            let x = Value::$Direct(x.into());
+            let child_width = x.width_or_child_width();
+            let address = self.buffer.len();
+            store_value(&mut self.buffer, x, child_width);
+            self.values.push(
+                Value::Reference {
+                    address,
+                    child_width,
+                    fxb_type: FlexBufferType::$Indirect,
+                }
+            );
+        }
+    }
+}
+#[derive(Debug, Clone)]
+struct CachedKey {
+    address: usize,
+    length: usize,
+}
+/// **Use this struct to build a Flexbuffer.**
+///
+/// Flexbuffers may only have a single root value, which may be constructed
+/// with  one of the following functions.
+/// * `build_singleton` will push 1 value to the buffer and serialize it as the root.
+/// * `build_vector` returns a `VectorBuilder`, into which many (potentially
+/// heterogenous) values can be pushed. The vector itself is the root and is serialized
+/// when the `VectorBuilder` is dropped (or `end` is called).
+/// * `build_map` returns a `MapBuilder`, which is similar to a `VectorBuilder` except
+/// every value must be pushed with an associated key. The map is serialized when the
+/// `MapBuilder` is dropped (or `end` is called).
+///
+/// These functions reset and overwrite the Builder which means, while there are no
+/// active `MapBuilder` or `VectorBuilder`, the internal buffer is empty or contains a
+/// finished Flexbuffer. The internal buffer is accessed with `view`.
+///
+/// The default builder interns keys. In the future, there should be string interning,
+/// and potentially key-vector interning.
+// TODO(cneo): Stablize options before 1.0.
+#[derive(Debug, Clone)]
+pub struct Builder {
+    buffer: Vec<u8>,
+    values: Vec<Value>,
+    key_pool: Option<Vec<CachedKey>>,
+}
+impl Default for Builder {
+    fn default() -> Self {
+        Builder {
+            key_pool: Some(Vec::new()),
+            values: Vec::new(),
+            buffer: Vec::new(),
+        }
+    }
+}
+
+
+impl<'a> Builder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+    /// Shows the internal flexbuffer. It will be empty on initialization or if `reset` was called.
+    /// Otherwise after `build_map`, `build_vector`, or `build_singleton` the internal flexbuffer
+    /// will be completed.
+    pub fn view(&self) -> &[u8] {
+        &self.buffer
+    }
+    /// Resets the internal state. Automatically called before building a new flexbuffer.
+    pub fn reset(&mut self) {
+        self.buffer.truncate(0);
+        self.values.truncate(0);
+        if let Some(pool) = self.key_pool.as_mut() {
+            pool.truncate(0);
+        }
+    }
+    fn push_key(&mut self, key: &str) {
+        assert!(
+            key.bytes().all(|b| b != b'\0'),
```

Comment:
Does this get turned into a memcmp by LLVM? @aardappel

Reply:
this isn't exactly a memcmp, since it appears to test that all bytes are not zero, for which there is no intrinsic, I believe. So this will likely result in an actual loop in code. But that is only in debug mode, right?

Reply:
@aardappel `assert!` always asserts; `debug_assert!` does it only in debug mode. I would guess that this assert is a guard in response to the recent flexbuffers key string problem.

Reply:
then we should probably be using `debug_assert`, as you wouldn't want to pay the cost of such checks in release builds.

And this has nothing to do with https://github.com/google/flatbuffers/issues/5627, this code is more general in that we want to use the Rust `str` type for convenience, even though the string being serialized is null-terminated, and thus can't represent embedded nulls.

Reply:
Made it debug_assert